requests-for-comments/2023-03-01-Temporal-RFC.md

```diff
@@ -0,0 +1,239 @@
+
+## Requirements
+
+We have a few products that are good candidates for a workflow framework. We’ll cover a few of the requirements here before we jump into any proposals just so that we can have a better understanding of what we are looking for.
+
+### Billing
+
+Billing is one of those things that is really tricky. Ideally you don’t want to bill someone twice (oof). You also don’t want to think you bill someone and then not bill them at all. Ideally you can kick off a flow and at the end of it have confidence that everything worked. The system was able to grab the usage metrics, apply the billing rate, and charge the user without having to double check the usage and the stripe api. Even though I still believe in trust, but verify. Any billing mistakes are generally pretty embarrassing.
+
+Requirements:
+
+- **Idempotent** or **exactly once** - no duplicates
+- **Retry-able** - If the thing fails, keep trying to get the money we are owed
+- **Resumable** - nice to pick up where you left off
+- **Composable** - Query ClickHouse. Calculate bill. Charge stripe.
+- **Parallelize-able** - We should be able to run multiple bills at the same time
+- **Timeouts** - We should be able to set timeouts on steps
+- **Cancelable** - We should be able to cancel any billing gracefully
+- **Terminate-able** - We should be able to rip the carpet out from a billing routine
+- **Nonblocking failures** - A failure of one billing workflow should not halt the execution of others.
+
+### Celery tasks
+
+Our celery tasks are pretty diverse. We use it for collecting metrics, refreshing caches, and if you didn’t know - we used to even use it to process every single event 😱.
+
+What do we use Celery for now?
+
+In general they are all scheduled tasks:
+
+- Running queries and emitting metrics (statsd, prometheus)
+- Warming up caches
+
+Requirements:
+
+- **Retry-able** - If the thing fails, get back on that horse
+- **Resumable** - Some of these are heavy and we don’t want to requery CH
+- **Parallelize-able** - We should be able to run multiple tasks at once
+- **Timeouts** - We should be able to set timeouts on the task
+- **Terminate-able** - We should be able to rip the carpet out from under the Task
+- **Nonblocking failures** - A failure of one task should not halt the execution of others.
+
+### Historical Exports *(AKA Vanilla ETL)*
+
+We do a lot of moving around data. Actually, a good chunk of our business is just getting data form A to B safe and sound, and in on piece. Some of these jobs are short, some are long running. There are a few requirements and nice to haves around ETLs
+
+Requirements:
+
+- **Idempotent** or **exactly once** - no duplicates
+- **Retry-able** - If the thing fails, get back on that horse
+- **Resumable** - Nice to pick up where you left off
+- **DAG-able** - Most ETLs resemble a Directed Acyclical Graph. Move A → B → C
+    - You can think of this as **Composable**
```

Comment:
A corollary of this requirement could be cycle-detection, in case we ever (accidentally) write a workflow that spins forever...

Reply:
Temporal doesn't do this just yet, you can basically just spawn child workflows (that are the same as the parent) forever. It's an intentional behaviour, though there has been an open issue on their github for ages and ages about being able to limit the stack depth there